8325859: Potential information loss during type inference

vicente-romero-oracle opened the pull request:
When an inference context (`IC`) is minimized, the minimized inference context (`MIC`) contains a proper subset of `IC's` inference vars (`IC_IV`). In other words there will be at least one inference variable `T` that belongs to `IC_IV` which doesn't belong to the minimized context inference variable's set (`MIC_IV`). MIC can be propagated to another inference context (AIC) as part of the type inference algo. If at any moment a type containing `T` needs, for example, to be instantiated `AIC` won't be able to perform that operation as it lacks the information to realize that `T` is a free variable it was supposed to know about hadn't a minimization step been performed. This PR is proposing to fix this issue without reducing the universe of inference contexts we apply the minimization algo to.

We are basically proposing to add a field to class InferenceContext which will be supplementary inference context that can be used to look for additional information if necessary. Also when registering a free type listener to an inference context, it could be that it won't be registered in the inference context the client could expect but in a supplementary inference context if it is found out that it contains and instantiation to all the free variables related to the listener. See also some comments inlined in the code,

TIA

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325859](https://bugs.openjdk.org/browse/JDK-8325859): Potential information loss during type inference (**Bug** - P3)


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25011/head:pull/25011` \
`$ git checkout pull/25011`

Update a local copy of the PR: \
`$ git checkout pull/25011` \
`$ git pull https://git.openjdk.org/jdk.git pull/25011/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25011`

View PR using the GUI difftool: \
`$ git pr show -t 25011`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25011.diff">https://git.openjdk.org/jdk/pull/25011.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25011#issuecomment-2847733366)
</details>

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java:
```diff
@@ -243,7 +262,72 @@ List<Type> instTypes() {
+        // no more type variables to be instantiated
+        if (next != null) {
+            HasTypeVars hasTypeVars = new HasTypeVars();
+            while (next != null && hasTypeVars.visit(result)) {
```
I think that we should be fine just going for the next supplementaryIC instead of traversing the whole list here. Same for the process to add free listeners, but in the first version I preferred to err on the correctness / completeness side. If we decide to visit more than one chained supplementary IC we can still do some improvements to try to reduce the number of supplementary ICs visited to a minimum

mcimadamore replied:
Maybe we can replace the visitor with asking `supplimentaryIC.free(t)` ? e.g. this could maybe use recusion?

```
Type result = t;
if (current.free(t)) {
    result = types.subst(t, inferencevars, instTypes());
}
if (supplimentaryIC != null) {
    return supplimentaryIC.asInstTypes(t);
 }
 ```

vicente-romero-oracle replied:
> Maybe we can replace the visitor with asking `supplimentaryIC.free(t)` ? e.g. this could maybe use recusion?
> 
> ```
> Type result = t;
> if (current.free(t)) {
>     result = types.subst(t, inferencevars, instTypes());
> }
> if (supplimentaryIC != null) {
>     return supplimentaryIC.asInstTypes(t);
>  }
> ```

right but won't this solution imply to recursively visit the whole chain of supplementaryICs? the current approach should stop as soon as there are no more supplementaryICs to visit. Actually I think that one of our performance tests failed and that was why I had to introduce this visitor if I recall correctly

mcimadamore replied:
> Actually I think that one of our performance tests failed and that was why I had to introduce this visitor if I recall correctly

I wanted to ask about this -- e.g. how much fixing this undoes the benefit of minimization. It seems we're in tricky territory.

Now, the problem with the visitor is that it picks up any type-variables -- it should at least filter out class type-variables and focus on method type-variables.

vicente-romero-oracle replied:
> > Actually I think that one of our performance tests failed and that was why I had to introduce this visitor if I recall correctly
> 
> I wanted to ask about this -- e.g. how much fixing this undoes the benefit of minimization. It seems we're in tricky territory.
> 
> Now, the problem with the visitor is that it picks up any type-variables -- it should at least filter out class type-variables and focus on method type-variables.

the difference in performance is negligible

mcimadamore commented:
The way this was designed, IIRC, was that if MIC is derived from IC, every time a variable is inferred in MIC, the resolution is forwarded to all the equivalent type variables in IC. If there's more type-variables in IC that are neither contained in MIC, nor equivalent to another var contained in MIC, it seems like an issue, as the minimization algo is not designed to work in this way.

liach commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java:
```diff
@@ -551,8 +651,18 @@ List<Type> solveBasic(List<Type> varsToSolve, EnumSet<InferenceStep> steps) {
+        String result = "Inference vars: " + inferencevars + '\n' +
+               "Undet vars: " + undetvars + '\n';
+        /*
+        // keep it for debugging
```
Should we control this with an `options.isSet` switch instead, maybe passing the options through Infer?

vicente-romero-oracle replied:
yes we could do that, but then we will probably want to add a test for that option, etc, probably not worthy, also we could just drop this code

vicente-romero-oracle commented:
> The way this was designed, IIRC, was that if MIC is derived from IC, every time a variable is inferred in MIC, the resolution is forwarded to all the equivalent type variables in IC. If there's more type-variables in IC that are neither contained in MIC, nor equivalent to another var contained in MIC, it seems like an issue, as the minimization algo is not designed to work in this way.

yes I think this is an issue of the minimization algo. I mean there will always be at least one variable in MIC that won't be in IC and if MIC is asked to substitute that variable by its corresponding inference variable, then MIC won't be able to do it and, potentially, some constraints won't be added to the inference variable. When we obtain MIC from IC we just don't know what queries we will do on MIC. The algo doesn't have this info.

I guess that there could be two ways to fix this:

- keeping links to the original inference contexts which will have more info, which is what this PR is doing
- trying to generate a more complete MIC gathering additional info

the second option seems like more complex as we could need some heuristics like type variables appearing in the same type, etc, and it could be harder to make sure that we get all the needed info. Also a more fluffy MIC would impact performance

mcimadamore commented:
> > The way this was designed, IIRC, was that if MIC is derived from IC, every time a variable is inferred in MIC, the resolution is forwarded to all the equivalent type variables in IC. If there's more type-variables in IC that are neither contained in MIC, nor equivalent to another var contained in MIC, it seems like an issue, as the minimization algo is not designed to work in this way.
> 
> yes I think this is an issue of the minimization algo. I mean there will always be at least one variable in MIC that won't be in IC and if MIC is asked to substitute that variable by its corresponding inference variable, then MIC won't be able to do it as it doesn't know how to map it and, potentially, some constraints won't be added to the inference variable. When we obtain MIC from IC we just don't know what queries we will do on MIC. The algo doesn't have this info.
> 
> I guess that there could be two ways to fix this:
> 
>     * keeping links to the original inference contexts which will have more info, which is what this PR is doing
> 
>     * trying to generate a more complete MIC gathering additional info
> 
> 
> the second option seems like more complex as we could need some heuristics like type variables appearing in the same type, etc, and it could be harder to make sure that we get all the needed info. Also a more fluffy MIC would impact performance

I agree that keeping a link to the original context is conceptually simpler -- I guess I'm worried that the implementation started off one way (e.g. avoid a link between context, and use free type listeners) and now it's moving in a different way (adding links between context). Which means that now we have both kinds of complexities.

More concretely, `InferenceContext::min` has code to track dependencies via listeners. This PR doesn't remove that code, but it adds a _new_ field, by which an inference context could point to its original version. Then it adds some more code to make sure that when we instantiate a type var on a context, all the contexts it depend on get the memo. While I believe this fixes the issue, I wouldn't be surprised if the original inference context would now be notified multiple times, once through the legacy mechanism, and once through the new mechanism.

vicente-romero-oracle commented:
> > > The way this was designed, IIRC, was that if MIC is derived from IC, every time a variable is inferred in MIC, the resolution is forwarded to all the equivalent type variables in IC. If there's more type-variables in IC that are neither contained in MIC, nor equivalent to another var contained in MIC, it seems like an issue, as the minimization algo is not designed to work in this way.
> > 
> > 
> > yes I think this is an issue of the minimization algo. I mean there will always be at least one variable in MIC that won't be in IC and if MIC is asked to substitute that variable by its corresponding inference variable, then MIC won't be able to do it as it doesn't know how to map it and, potentially, some constraints won't be added to the inference variable. When we obtain MIC from IC we just don't know what queries we will do on MIC. The algo doesn't have this info.
> > I guess that there could be two ways to fix this:
> > ```
> > * keeping links to the original inference contexts which will have more info, which is what this PR is doing
> > 
> > * trying to generate a more complete MIC gathering additional info
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the second option seems like more complex as we could need some heuristics like type variables appearing in the same type, etc, and it could be harder to make sure that we get all the needed info. Also a more fluffy MIC would impact performance
> 
> I agree that keeping a link to the original context is conceptually simpler -- I guess I'm worried that the implementation started off one way (e.g. avoid a link between context, and use free type listeners) and now it's moving in a different way (adding links between context). Which means that now we have both kinds of complexities.
> 
> More concretely, `InferenceContext::min` has code to track dependencies via listeners. This PR doesn't remove that code, but it adds a _new_ field, by which an inference context could point to its original version. Then it adds some more code to make sure that when we instantiate a type var on a context, all the contexts it depend on get the memo. While I believe this fixes the issue, I wouldn't be surprised if the original inference context would now be notified multiple times, once through the legacy mechanism, and once through the new mechanism.
> 
> Is it possible to simplify the code so that the old mechanism goes away, and we do everything using the new field?

I have made several experiments to avoid the use of the free type listeners, the prototypes are just moving the code in the listeners to other places basically keeping the same level of complexity but trying to remove the timing component associated to the listeners. I have fixed several bugs or missing parts in the prototypes but the build then fails somewhere else. Also above you said:

>Then it adds some more code to make sure that when we instantiate a type var on a context, all the contexts it depend on get the memo

there is no cascade notification to other contexts when a type var is instantiated, the new code is just using the supplementary contexts as a kind of dictionary to look for the inference vars a given inference context doesn't know about. Given that there are no cascade notifications I don't think that the original inference context will be notified multiple times. Anyways multiple notification should not be possible as every time a free type listener is executed it is removed from the free type listener's map.

mcimadamore commented:
Btw, I've looked at the issue in JBS again, and simplified it a bit:

```
class Generics {
       public static void main(String... args) {
            Integer i = null;
            runT(() -> supplyNull(i));
        }

        static <R, X> R supplyNull(X... varargs) {
            System.out.println(varargs.getClass().getComponentType());
            return null;
        }
              
        static <T> void runT(Runnable runnable) { runnable.run(); }
    //}
```

(I've removed a call to `Integer::valueOf` that was not necessary to reproduce the issue -- but was adding some extra noise).

Note that here the calls to `supplyNull` and `runT` are completely independent. That's because the call to `supplyNull` occurs in a void-compatible lambda (the target of the lambda is `Runnable`). As such, it is not possible for inference of the nested call to `supplyNull` to have any effect on the inference of the `runT` call.

However, when debugging, I see that when we call `checkMethod` for `supplyNull`, we attempt to propagate its inference context outwards. This is done even though the `resultInfo` says that the expected type is `void` (where `void` is the return type of the function descriptor).

In other words, it seems to me that `Infer::shouldPropagate` is buggy: it returns true if all the conditions below are satisfied:
* the enclosing method is a generic method
* the return type of the current method contains some inference vars
* no eager resolution is required

But we never check what the target-type of the current method actually is. Normally, the target will be some non-void type -- so when we propagate, we will need to check compatibility again, which will fix up all the types. But in this case (which is a corner case due to void-compatible lambdas), the target happens to be `void` -- meaning inference will never be "completed".

See the implementation of `Attr.ExpressionLambdaReturnContext::compatible`:

```
           @Override
            public boolean compatible(Type found, Type req, Warner warn) {
                //a void return is compatible with an expression statement lambda
                if (req.hasTag(VOID)) {
                    expStmtExpected = true;
                    return TreeInfo.isExpressionStatement(expr);
                } else {
                    return super.compatible(found, req, warn);
                }
            }
```

In other words, if the expected type is `void` (as in this case), `super.compatible` will not be called -- which means the deferred type for the nested generic method call will remain uninferred.

Adding an extra check to `Infer::shouldPropagate` to _not_ propagate when the target happens to be `void` seems to fix the issue in JBS. (but, I have not ran any other test).

vicente-romero-oracle commented:
> Btw, I've looked at the issue in JBS again, and simplified it a bit:
> 
> ```
> class Generics {
>        public static void main(String... args) {
>             Integer i = null;
>             runT(() -> supplyNull(i));
>         }
> 
>         static <R, X> R supplyNull(X... varargs) {
>             System.out.println(varargs.getClass().getComponentType());
>             return null;
>         }
>               
>         static <T> void runT(Runnable runnable) { runnable.run(); }
> }
> ```
> 
> (I've removed a call to `Integer::valueOf` that was not necessary to reproduce the issue -- but was adding some extra noise).
> 
> Note that here the calls to `supplyNull` and `runT` are completely independent. That's because the call to `supplyNull` occurs in a void-compatible lambda (the target of the lambda is `Runnable`). As such, it is not possible for inference of the nested call to `supplyNull` to have any effect on the inference of the `runT` call.
> 
> However, when debugging, I see that when we call `checkMethod` for `supplyNull`, we attempt to propagate its inference context outwards. This is done even though the `resultInfo` says that the expected type is `void` (where `void` is the return type of the function descriptor).
> 
> In other words, it seems to me that `Infer::shouldPropagate` is buggy: it returns true if all the conditions below are satisfied:
> 
>     * the enclosing method is a generic method
> 
>     * the return type of the current method contains some inference vars
> 
>     * no eager resolution is required
> 
> 
> But we never check what the target-type of the current method actually is. Normally, the target will be some non-void type -- so when we propagate, we will need to check compatibility again, which will fix up all the types. But in this case (which is a corner case due to void-compatible lambdas), the target happens to be `void` -- meaning inference will never be "completed".
> 
> See the implementation of `Attr.ExpressionLambdaReturnContext::compatible`:
> 
> ```
>            @Override
>             public boolean compatible(Type found, Type req, Warner warn) {
>                 //a void return is compatible with an expression statement lambda
>                 if (req.hasTag(VOID)) {
>                     expStmtExpected = true;
>                     return TreeInfo.isExpressionStatement(expr);
>                 } else {
>                     return super.compatible(found, req, warn);
>                 }
>             }
> ```
> 
> In other words, if the expected type is `void` (as in this case), `super.compatible` will not be called -- which means the deferred type for the nested generic method call will remain uninferred.
> 
> Adding an extra check to `Infer::shouldPropagate` to _not_ propagate when the target happens to be `void` seems to fix the issue in JBS. (but, I have not ran any other test).

good catch, yes this is a bug in `Infer::shouldPropagate` and one we should fix for sure, but I think that the fundamental problem is still there and will show its head again as soon as, for example, the lambda is not void-compatible

mcimadamore commented:
> good catch, yes this is a bug in `Infer::shouldPropagate` and one we should fix for sure, but I think that the fundamental problem is still there and will show its head again as soon as, for example, the lambda is not void-compatible

If the target type is non-void, then it will be checked by the `compatible` method I shown above. This is a deferred type check which will cause the inference context to be inferred in full. The problem of this issue is that the final inference step is never executed -- because the target type is `void`.

If you can find a case w/o `void` that exhibits the same issue, we can certainly look into that more.

vicente-romero-oracle commented:
> > good catch, yes this is a bug in `Infer::shouldPropagate` and one we should fix for sure, but I think that the fundamental problem is still there and will show its head again as soon as, for example, the lambda is not void-compatible
> 
> If the target type is non-void, then it will be checked by the `compatible` method I shown above. This is a deferred type check which will cause the inference context to be inferred in full. The problem of this issue is that the final inference step is never executed -- because the target type is `void`.
> 
> If you can find a case w/o `void` that exhibits the same issue, we can certainly look into that more.

I'm seeing the same issue as the one reported with this test case:

```
import java.util.function.*;

class Generic {

    public static void main(String... args) {
        runT(() -> supplyNull(Integer.valueOf(1)));
    }

    static <R, X> R supplyNull(X... varargs) {
        System.out.println(varargs.getClass().getComponentType());
        return null;
    }

    static <T> void runT(Supplier<T> supp) { supp.get(); }
}
```

vicente-romero-oracle commented:
> > If you can find a case w/o `void` that exhibits the same issue, we can certainly look into that more.
> 
> Uhmmm. This:
> 
> ```
> static class Generic {
> 
>         public static void main(String... args) {
>             Integer i = null;
>             runT(() -> supplyNull(i));
>         }
> 
>         static <R, X> R supplyNull(X... varargs) {
>             System.out.println(varargs.getClass().getComponentType());
>             return null;
>         }
> 
>         static <T> void runT(Supplier<T> runnable) { runnable.get(); }
>     }
> ```
> 
> Seems to exhibit same incorrect behavior. Inference goes a bit further (I see R being replaced, but not X).

right I was writing a very similar example, and X is not replaced because the minimized inference context doesn't know about X

mcimadamore commented:
> right I was writing a very similar example, and X is not replaced because the minimized inference context doesn't know about X

That is true -- however, when we create the minimized context, we set a listener from the minimized context to the original context, so that whenever *any* propagated variable is solved, we also solve the "independent" variables. This listener works -- the problem is that the replacement in the inferred method signature never occurs. So I'd like to spend some more time understanding exactly why this happens.

mcimadamore commented:
> > right I was writing a very similar example, and X is not replaced because the minimized inference context doesn't know about X
> 
> That is true -- however, when we create the minimized context, we set a listener from the minimized context to the original context, so that whenever _any_ propagated variable is solved, we also solve the "independent" variables. This listener works -- the problem is that the replacement in the inferred method signature never occurs. So I'd like to spend some more time understanding exactly why this happens.

It also seems to me that this problem has been there for a long time -- possibly since the introduction of `InferenceContext::min` -- we do all the correct checks, but some of the types are left "uninferred" in the AST. Which then shows up whenever the translation strategy relies on the type being inferred (such as here -- as we need the inferred type to know which vararg array to create).

mcimadamore commented:
I think I see the problem now. I believe the issue is in `Check::checkMethod`. That method does this at the start:

```
if (inferenceContext.free(mtype)) {
            inferenceContext.addFreeTypeListener(List.of(mtype),
                    solvedContext -> checkMethod(solvedContext.asInstType(mtype), sym, env, argtrees, argtypes, useVarargs, solvedContext));
            return mtype;
        }
```

This is correct -- if `mtype` has some free type variable, we need to come back later and re-run this again (to set the correct type on the AST).

But there's an issue -- if we propagate (`Infer::shouldPropagate` is `true`), then the input inference context is NOT the inference context for the call being checked (in this case `supplyNull`) but the outer context, which contains the propagated vars.

While this was correct before the introduction of context minimization -- because in such case outer inference contexts were always strict supersets of nested inference contexts "contained", this is no longer true with minimization: we can have an outer context that omits some variables from the nested context. At which point, the listener above will "miss the memo" -- e.g. it will instantiate the method type using the outer context inferred vars -- but this will still leave "holes" (because the outer context will not contain all the variables that can appear in `mtype`).

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java:
```diff
@@ -83,6 +94,14 @@ Type update(Type t) {
+     * check for set membership for cases where the minimized context or any other context derived from it
+     * needs to deal with an inference variable that has been eliminated from IC_IV while minimizing it
+     */
+    InferenceContext supplementaryIC;
```
Let's call this `next`, or `parent` ?

vicente-romero-oracle replied:
I think it is more a `child` context than a `parent` one, `nested`?

vicente-romero-oracle replied:
I see, yes makes sense

vicente-romero-oracle commented:
> I think I see the problem now. I believe the issue is in `Check::checkMethod`. That method does this at the start:
> 
> ```
> if (inferenceContext.free(mtype)) {
>             inferenceContext.addFreeTypeListener(List.of(mtype),
>                     solvedContext -> checkMethod(solvedContext.asInstType(mtype), sym, env, argtrees, argtypes, useVarargs, solvedContext));
>             return mtype;
>         }
> ```
> 
> This is correct -- if `mtype` has some free type variable, we need to come back later and re-run this again (to set the correct type on the AST).
> 
> But there's an issue -- if we propagate (`Infer::shouldPropagate` is `true`), then the input inference context is NOT the inference context for the call being checked (in this case `supplyNull`) but the outer context, which contains the propagated vars.
> 
> While this was correct before the introduction of context minimization -- because in such case outer inference contexts were always strict supersets of nested inference contexts "contained", this is no longer true with minimization: we can have an outer context that omits some variables from the nested context. At which point, the listener above will "miss the memo" -- e.g. it will instantiate the method type using the outer context inferred vars -- but this will still leave "holes" (because the outer context will not contain all the variables that can appear in `mtype`).

correct, these were my findings too, and if we try to make the minimized context more `chubby` then for a lot of cases we will probably kill the whole point of doing the minimization at all. So the supplementary context was the solution I found to keep the whole minimization algo going while doing some bookkeeping in case it could be needed by an outer inference context

mcimadamore commented:
> Another option, but I think it will be more complex, could be for the minimization algo to don't physically reduce the size of a given inference context but to make some variables "inactive" so that they won't participate in the propagation and resolution algos but they can be queried if needed

Yes, that would be another way to do it -- e.g. keep the outer inference context a strict superset of the nested one -- but make incorporation only touch few bits.

Which... leads me to a possible "interim" solution: when we create the min context, we know which variables are NOT going to be propagated out. Is there a way we can somehow save these variables on the outer context -- in  separate lists? If we could do that, then `asInstTypes` only needs a small tweak to use the non-minimized variable list instead of the minimized one.

In a way we can decide whether to save a pointer to the original context, or whether to store all the actual vars (but in a separate place, so it won't be used by the other inference machinery). Maybe we should try both approaches and see which one leads to simpler code?

vicente-romero-oracle commented:
> > Another option, but I think it will be more complex, could be for the minimization algo to don't physically reduce the size of a given inference context but to make some variables "inactive" so that they won't participate in the propagation and resolution algos but they can be queried if needed
> 
> Yes, that would be another way to do it -- e.g. keep the outer inference context a strict superset of the nested one -- but make incorporation only touch few bits.
> 
> Which... leads me to a possible "interim" solution: when we create the min context, we know which variables are NOT going to be propagated out. Is there a way we can somehow save these variables on the outer context -- in separate lists? If we could do that, then `asInstTypes` only needs a small tweak to use the non-minimized variable list instead of the minimized one.
> 
> In a way we can decide whether to save a pointer to the original context, or whether to store all the actual vars (but in a separate place, so it won't be used by the other inference machinery). Maybe we should try both approaches and see which one leads to simpler code?

yes this is exactly the way I thought we could approach this other alternative, using different lists for the variables we don't want to participate in the outer context resolution and propagation steps

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java:
```diff
@@ -258,7 +276,23 @@ List<Type> asInstTypes(List<Type> ts) {
     void addFreeTypeListener(List<Type> types, FreeTypeListener ftl) {
-        freeTypeListeners.put(ftl, freeVarsIn(types));
+        List<Type> myFreeVars = freeVarsIn(types);
+        if (parentIC == null) {
```
I'm not sure I get this. In the sense: I get what the code is doing, but I'm not sure I understand why it's needed. E.g. the problem we're trying to solve is when an outer method has some inference variables that have not been propagated. This means that when the outer inference context is solved, calling `asInstTypes` on that outer context won't replace all the types. But your new patch to `asInstTypes` collects (correctly) all the inference variables from all parent contexts -- and uses those to do the type subst. Isn't that enough?

I'm a bit surprised that we also need to be picky when adding listeners -- as I'd assume that once the outer inference context is solved, all the nested inference contexts will also be solved too?

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java:
```diff
@@ -552,7 +570,7 @@ List<Type> solveBasic(List<Type> varsToSolve, EnumSet<InferenceStep> steps) {
     public String toString() {
         return "Inference vars: " + inferencevars + '\n' +
-               "Undet vars: " + undetvars;
+               "Undet vars: " + undetvars + '\n';
```
Optional: should this also print the parent?

vicente-romero-oracle replied:
sure, done

mcimadamore reviewed (approved):
The new simpler patch looks great! Thanks for persevering!!!

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java:
```diff
@@ -551,8 +569,12 @@ List<Type> solveBasic(List<Type> varsToSolve, EnumSet<InferenceStep> steps) {
+        String result = "Inference vars: " + inferencevars + '\n' +
+               "Undet vars: " + undetvars + '\n';
+        if (parentIC != null) {
+            result += "\nParent : " + parentIC.toString();
```
I was thinking about using an indentation, probably with a toString overload, but if the chain of parents gets long then one will need to scroll to see them all

mcimadamore replied:
I see what you mean. The fact that the string representation is multiline, but also recursively used leads to issues -- because every time we do a `\n` we will break indentation. It's not terribly important. Another possibility would be to list the "other variables" (e.g. the ones that appear in the parent contexts), without listing all the parent contexts explicitly. That should scale a bit better.

vicente-romero-oracle replied:
sorry you mean showing only the variables of the next parent context and ignore the rest of the chain?

mcimadamore replied:
> you mean showing only the variables of the adjacent parent context and ignoring the rest of the chain?

Yeah -- but perhaps that's not very helpful as some of those variables will have same name even if different... Just leave it for now :-)

vicente-romero-oracle replied:
right, ship sailing, all tests are green, thanks!

vicente-romero-oracle commented:
/integrate
